File: libs/power-log-parser/src/lib/parsers/shuffle-deck-parser.ts
import { ActionParser } from '../action-parser';
import { GameEventProvider } from '../game-event';
import { Node, ShuffleDeck } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class ShuffleDeckParser implements ActionParser {
	readonly ParserName = 'ShuffleDeckParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		return stateType === StateType.PowerTaskList && node.Type === ShuffleDeck;
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const shuffleDeck = node.Object as ShuffleDeck;
		return [
			GameEventProvider.Create(
				shuffleDeck.TimeStamp,
				'SHUFFLE_DECK',
				() => ({
					Type: 'SHUFFLE_DECK',
					Value: {
						PlayerId: shuffleDeck.PlayerId,
						LocalPlayer: this.StateFacade.LocalPlayer,
						OpponentPlayer: this.StateFacade.OpponentPlayer,
					},
				}),
				false,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/ShuffleDeckParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using System.Linq;

namespace HearthstoneReplays.Events.Parsers
{
    public class ShuffleDeckParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public ShuffleDeckParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(ShuffleDeck);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var shuffleDeck = node.Object as ShuffleDeck;
            return new List<GameEventProvider>()
            {
                GameEventProvider.Create(
                   shuffleDeck.TimeStamp,
                   "SHUFFLE_DECK",
                   () =>
                   {
                       return new GameEvent
                       {
                           Type = "SHUFFLE_DECK",
                           Value = new
                           {
                               PlayerId = shuffleDeck.PlayerId,
                               LocalPlayer = StateFacade.LocalPlayer,
                               OpponentPlayer = StateFacade.OpponentPlayer,
                           }
                       };
                   },
                   false,
                   node)
            };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
